### PR TITLE
IElementStateProvider tests

### LINKED ISSUE #7: Add tests for IElementStateProvider (State property at IElement)

### PR DESCRIPTION
**Pull request by hsuboch**
closes #7

### CODE REVIEW

File: Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        [Test]
+        public void Should_ReturnTrue_IfElementIsEnabled()
+        {
+            Assert.IsTrue(RightArgumentTextBox.State.IsEnabled);
+        }
+
+        [Test]
+        public void Should_ThrowException_IfWaitForClickableEnded()
+        {
+            Assert.Throws<WebDriverTimeoutException>(() => EmptyButton.State.WaitForClickable(TimeSpan.Zero));
```
Comment (DmitryBogatko) [inline Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs:83]: @hsuboch , @mialeska don't we want to check that wait lasts expected time? The same question for all wait methods. Or will it be covered or has been already covered in core tests?

**Review by DmitryBogatko — DISMISSED** [review]

File: Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs (hunk 1 above)
Comment (mialeska) [reply Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs:83]: yes, I believe we should add such tests

File: Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        [Test]
+        public void Should_ReturnFalse_IfElementIsNotEnabled()
+        {
+            Assert.IsFalse(EmptyButton.State.IsEnabled);
+        }
+
+        [Test]
+        public void Should_ReturnTrue_IfElementIsEnabled()
+        {
+            Assert.IsTrue(RightArgumentTextBox.State.IsEnabled);
```
Comment (mialeska) [inline Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs:77]: Seems that you need to update nuget package Aquality,Selenium.Core to make this test fixed

Comment (mialeska) [reply Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs:77]: please ensure that all tests are passed

**Review by mialeska — CHANGES_REQUESTED** [review]

File: Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs (hunk 2 above)
Comment (hsuboch) [reply Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs:77]: done

File: Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs (hunk 1 above)
Comment (hsuboch) [reply Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs:83]: added

File: Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+        [Test]
+        public void Should_WaitForNotEnabled_CorrectAmountOfTime()
+        {
+            var stopwatch = StartedStopwatch;
+            RightArgumentTextBox.State.WaitForNotEnabled(FromSeconds);
+            AssertIfWaitTimeIsCorrect(stopwatch);
+        }
+
+        private static void AssertIfWaitTimeIsCorrect(Stopwatch stopwatch)
```
Comment (mialeska) [inline Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs]: seems that *_CorrectAmountOfTime tests have similar actions and could be simplified with the following signature:

using IElementStateProvider = Aquality.Selenium.Core.Elements.Interfaces.IElementStateProvider;


        private static void AssertIfWaitTimeIsCorrect(IElementStateProvider stateProvider, Action<IElementStateProvider, TimeSpan> action)
        {
            var stopwatch = StartedStopwatch;
            action(stateProvider, FromSeconds);
            stopwatch.Stop();
            Assert.LessOrEqual(Math.Abs(stopwatch.Elapsed.Seconds - FromSeconds.Seconds), 1, "Wait Time is NOT correct");
        }
`

        [Test]
        public void Should_WaitForNotEnabled_CorrectAmountOfTime()
        {
            AssertIfWaitTimeIsCorrect(RightArgumentTextBox.State, (state, time) => state.WaitForNotEnabled(time));
        }

        [Test]
        public void Should_WaitForClickable_CorrectAmountOfTime()
        {
            AssertIfWaitTimeIsCorrect(EmptyButton.State, (state, time) =>
            {
                try
                {
                    EmptyButton.State.WaitForClickable(FromSeconds);
                }
                catch (WebDriverTimeoutException)
                {
                }
            });
        }

File: Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+{
+    public class ElementStateProviderTests : TestWithApplication
+    {
+        private static readonly CalculatorWindow CalculatorWindow = new CalculatorWindow();
+        private static readonly ITextBox RightArgumentTextBox = CalculatorWindow.RightArgumentTextBox;
+        private static readonly IButton MaximizeButton = CalculatorWindow.MaximizeButton;
+        private static readonly IButton EmptyButton = CalculatorWindow.EmptyButton;
+        private static readonly TimeSpan FromSeconds = TimeSpan.FromSeconds(5);
+
+        private IElement notPresentLabel = ApplicationManager.GetRequiredService<IElementFactory>()
```
Comment (mialeska) [inline Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs]: should be readonly

**Review by mialeska — CHANGES_REQUESTED** [review]

File: Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs (hunk 2 above)
Comment (mialeska) [reply Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs:77]: failing again

Comment (hsuboch) [reply Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs:77]: fixed

File: Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs (hunk 3 above)
Comment (hsuboch) [reply Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs]: done

File: Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs (hunk 4 above)
Comment (hsuboch) [reply Aquality.WinAppDriver/tests/Aquality.WinAppDriver.Tests/Elements/ElementStateProviderTests.cs]: done

**Review by mialeska — APPROVED** [review]

**Pull request merged**